File: algorithm-java/src/main/java/com/algorithm/chaptertwo/sort/third/QuickSort.java
package com.algorithm.chaptertwo.sort.third;

import com.algorithm.chaptertwo.sort.util.SortUtils;
import edu.princeton.cs.algs4.StdRandom;

/**
 * @author clx 2/17/2018
 */
public class QuickSort {

	/**
	 * quick sort
	 *
	 * @param array
	 */
	public static void sort(Integer[] array) {
		if (SortUtils.checkArrayValid(array)) {
			return;
		}
		// shuffle
		StdRandom.shuffle(array);
		QuickSort.sort(array, 0, array.length - 1);
	}

	/**
	 * sort
	 *
	 * @param array
	 * @param lo
	 * @param hi
	 */
	private static void sort(Integer[] array, int lo, int hi) {
		if (hi <= lo) {
			return;
		}
		// partition array
		int idx = QuickSort.partition(array, lo, hi);
		// sort left
		sort(array, lo, idx - 1);
		// sort right
		sort(array, idx + 1, hi);
	}

	/**
	 * partition array
	 *
	 * @param array
	 * @param lo
	 * @param hi
	 */
	private static int partition(Integer[] array, int lo, int hi) {
		int idx = lo;
		int jdx = hi + 1;
		Integer pivot = array[lo];
		while (true) {
			while (SortUtils.less(array[++idx], pivot)) {
				if (idx == hi) {
					break;
				}
			}
			while (SortUtils.less(pivot, array[--jdx])) {
				if (jdx == lo) {
					break;
				}
			}
			if (idx >= jdx) {
				break;
			}
			SortUtils.exchange(array, idx, jdx);
		}
		SortUtils.exchange(array, lo, jdx);
		return jdx;
	}
}

File: algorithm-go/src/grokkingalgorithms/forthchapter/quick_sort.go
package forthchapter

import "grokkingalgorithms/common"

// QuickSort sort array
func QuickSort(array []int) []int {
	if array == nil {
		return nil
	}
	if len(array) == 0 {
		return nil
	}
	sort(array, 0, len(array)-1)
	return array
}

func sort(array []int, lo, hi int) {
	if hi <= lo {
		return
	}
	// partition array
	idx := partition(array, lo, hi)
	// sort left
	sort(array, lo, idx-1)
	// sort right
	sort(array, idx+1, hi)
}

func partition(array []int, lo, hi int) int {
	idx := lo + 1
	jdx := hi
	pivot := array[lo]
	for true {
		for ; common.Less(array[idx], pivot); idx++ {
			if idx == hi {
				break
			}
		}
		for ; common.Less(pivot, array[jdx]); jdx-- {
			if jdx == lo {
				break
			}
		}
		if idx >= jdx {
			break
		}
		common.Exchange(array, idx, jdx)
	}
	common.Exchange(array, lo, jdx)
	return jdx
}
